ballista/scheduler/src/state/mod.rs

```diff
@@ -236,46 +236,53 @@ impl<T: 'static + AsLogicalPlan, U: 'static + AsExecutionPlan> SchedulerState<T,
-                                    tasks,
-                                    &self.executor_manager,
-                                )
-                                .await
-                            {
-                                error!("Failed to launch new task: {:?}", e);
-                                for _i in 0..n_tasks {
-                                    unassigned_reservations.push(
-                                        ExecutorReservation::new_free(
-                                            executor_id.clone(),
-                                        ),
-                                    );
+                    let task_manager = self.task_manager.clone();
+                    let executor_manager = self.executor_manager.clone();
+                    let join_handle = tokio::spawn(async move {
+                        let success = match executor_manager
+                            .get_executor_metadata(&executor_id)
+                            .await
+                        {
+                            Ok(executor) => {
+                                if let Err(e) = task_manager
+                                    .launch_multi_task(
+                                        &executor,
+                                        tasks,
+                                        &executor_manager,
+                                    )
+                                    .await
+                                {
+                                    error!("Failed to launch new task: {:?}", e);
+                                    false
+                                } else {
+                                    true
                                 }
                             }
-                        }
-                        Err(e) => {
-                            error!("Failed to launch new task, could not get executor metadata: {:?}", e);
-                            for _i in 0..n_tasks {
-                                unassigned_reservations.push(
-                                    ExecutorReservation::new_free(executor_id.clone()),
-                                );
+                            Err(e) => {
+                                error!("Failed to launch new task, could not get executor metadata: {:?}", e);
+                                false
                             }
+                        };
+                        if success {
+                            vec![]
+                        } else {
+                            vec![
+                                ExecutorReservation::new_free(executor_id.clone(),);
+                                n_tasks
+                            ]
                         }
-                    }
+                    });
+                    join_handles.push(join_handle);
+                }
+                for join_handle in join_handles {
+                    unassigned_reservations.append(&mut join_handle.await?);
```

Comment:
Not sure I understand, wouldn't this still process the work sequentially? I would we would want do do a `futures::future::join_all` on the collection of futures to actually execute them in parallel

Reply:
Hi @thinkharderdev, thanks for your advice. The blocking part is calling the method _launch_multi_task_ which has already been invoked in parallel. However, `futures::future::join_all` is a much more simple way. I'll refine it.